### PR TITLE
Add decrement operator for ArrayIterator

### PR DESCRIPTION
**Pull request by wbernoudy**

### CODE REVIEW

**Comment by arcondello** [comment]
We should add another static assert [here](https://github.com/dwavesystems/dwave-optimization/blob/a627daf9d5f3007012df9094ba0dd04aa025e6d2/dwave/optimization/include/dwave-optimization/array.hpp#L470) for [bidirectional_iterator](https://en.cppreference.com/w/cpp/iterator/bidirectional_iterator).

File: dwave/optimization/include/dwave-optimization/array.hpp [hunk 1]
```diff
@@ -413,17 +435,40 @@ class ArrayIterator {
...
+                for (std::ptrdiff_t i = 0; i > n; --i) {
+                    offset += unadvance();
+                }
+            } else {
+                for (std::ptrdiff_t i = 0; i < n; ++i) {
+                    offset += advance();
+                }
             }
             return offset;
         }
```
Comment (arcondello) [inline dwave/optimization/include/dwave-optimization/array.hpp:474]: I think we could also just do
```c++
        std::ptrdiff_t advance(std::ptrdiff_t n) {
            std::ptrdiff_t offset = 0;

            for (; n > 0; --n) offset += unadvance();
            for (; n < 0; ++n) offset += advance();

            return offset;
        }
```
likely after the compiler has at it, it won't make much difference. But I guess this is a few fewer instructions/memory locations if it's taken literally.

**Review by arcondello — APPROVED** [review]
LGTM!

File: dwave/optimization/include/dwave-optimization/array.hpp (hunk 1 above)
Comment (wbernoudy) [reply dwave/optimization/include/dwave-optimization/array.hpp]: Done, though I had to switch them :stuck_out_tongue:

**Pull request merged**